cpp/include/cudf/join.hpp

```diff
@@ -438,6 +446,65 @@ class hash_join {
   const std::unique_ptr<impl_type const> _impl;
 };
 
+/**
+ * @brief Unique hash join that builds hash table in creation and probes results in subsequent
+ * `*_join` member functions
+ *
+ * @note Behavior is undefined if the build table contains duplicates.
+ * @note All NaNs are considered as equal
+ *
+ * @tparam HasNested Flag indicating whether there are nested columns in build/probe table
+ */
+// TODO: `HasNested` to be removed via dispatching
+template <has_nested HasNested = has_nested::NO>
+class unique_hash_join {
+ public:
+  unique_hash_join() = delete;
+  ~unique_hash_join();
+  unique_hash_join(unique_hash_join const&)            = delete;
+  unique_hash_join(unique_hash_join&&)                 = delete;
+  unique_hash_join& operator=(unique_hash_join const&) = delete;
+  unique_hash_join& operator=(unique_hash_join&&)      = delete;
+
+  /**
+   * @brief Constructs a unique hash join object for subsequent probe calls
+   *
+   * @param build The build table that contains unique elements
+   * @param probe The probe table, from which the keys are probed
+   * @param has_nulls Flag to indicate if there exists any nulls in the `build` table or
+   *        any `probe` table that will be used later for join
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  unique_hash_join(cudf::table_view const& build,
+                   cudf::table_view const& probe,
+                   nullable_join has_nulls      = nullable_join::YES,
+                   null_equality compare_nulls  = null_equality::EQUAL,
+                   rmm::cuda_stream_view stream = cudf::get_default_stream());
+
+  /**
+   * Returns the row indices that can be used to construct the result of performing
+   * an inner join between two tables. @see cudf::inner_join().
+   *
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used to allocate the returned table and columns' device
+   * memory.
```

Comment:
This returns two uvectors of indices, not a table/columns.
```suggestion
   * @param mr Device memory resource used to allocate the returned indices' device memory.
```